### PR TITLE
feat(msr, persistence)!: refactor msr

### PR DESCRIPTION
**Pull request by mikevoronov**

### CODE REVIEW

File: crates/msr/src/lib.rs [hunk 1]
```diff
@@ -14,31 +14,31 @@
...
+pub use msr_impl::*;
 
 pub type MSRResult<T> = Result<T, MSRError>;
 
-pub trait MSR {
-    fn write_preset(&mut self, store_state: bool) -> MSRResult<()>;
-    fn repin(&mut self, core_id: LogicalCoreId) -> MSRResult<()>;
-    fn restore(self) -> MSRResult<()>;
+pub trait MSREnforce {
+    /// Applies chosen MSR policy to current core.
```
Comment (raftedproc) [inline crates/msr/src/lib.rs:39]: Preset, not policy.

File: crates/msr/src/linux_x86_64/msr_x86_64.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    is_enabled: bool,
+    original_preset: MSRCpuPreset,
+}
+
+impl MSRModeEnforcer {
+    /// Initialize enforces by reading the original values from OS,
+    /// this method should typically be called only once in CCP.
+    pub fn from_os(is_enabled: bool) -> Self {
+        let mode = MSRMode::detect();
+        let preset = mode.get_optimal_cpu_preset();
```
Comment (raftedproc) [inline crates/msr/src/linux_x86_64/msr_x86_64.rs:39]: You don't need to detect/save preset if it is disabled.

File: crates/msr/src/linux_x86_64/msr_x86_64.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+pub fn write(item: MSRPresetItem, core_id: LogicalCoreId) -> MSRResult<()> {
+    let value_to_write = if item.mask() != MSRPresetItem::NO_MASK {
+        let old_value = utils::read_msr(item.register_id(), core_id)?;
+        MSRPresetItem::masked_value(old_value, item.value(), item.mask())
+    } else {
+        item.value()
+    };
+
+    tracing::debug!(
+        "Write MSR register_id {:?} value {:} at logical CPU {}  ",
```
Comment (raftedproc) [inline crates/msr/src/linux_x86_64/msr_x86_64.rs:110]: Here is better to use hex for values.

File: crates/msr/src/linux_x86_64/utils.rs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+    let file = open_msr(core_id, MSRFileOpMode::MSRRead)
+        .map_err(|error| MSRError::open_for_read(core_id, error))?;
+
+    let mut value = [0u8; 8];
+    pread(file, &mut value, register_id as i64)
+        .map_err(|errno| MSRError::read_w_no_err(register_id, core_id, errno))?;
+    let result = u64::from_le_bytes(value);
+
+    tracing::debug!("Read MSR register_id {register_id:?} value {value:?} at core id {core_id}",);
```
Comment (raftedproc) [inline crates/msr/src/linux_x86_64/utils.rs:41]: Use hex printing the values out.

File: crates/shared/src/types/epoch_parameters.rs [hunk 5]
```diff
@@ -36,6 +37,15 @@ impl EpochParameters {
     }
 }
 
+impl From<CCProofId> for EpochParameters {
```
Comment (raftedproc) [inline crates/shared/src/types/epoch_parameters.rs:40]: How do you use this change in the patch b/c this looks unrelated.

File: ccp/src/proof_storage.rs [hunk 6]
```diff
@@ -81,14 +81,9 @@ impl ProofStorageDrainer {
 
     pub async fn validate_proofs(
         &mut self,
-        epoch_params: Option<&EpochParameters>,
+        epoch: &Option<EpochParameters>,
```
Comment (raftedproc) [inline ccp/src/proof_storage.rs:84]: This looks unrelated.

File: ccp/src/prover.rs [hunk 7]
```diff
@@ -109,31 +112,69 @@ impl ToCCStatus for CCProver {
...
+    pub async fn from_saved_state(config: CCPConfig) -> CCResult<Self> {
+        let state_storage = StateStorage::new(config.state_dir.clone());
+        let prev_state = match state_storage.try_to_load_data().await? {
+            Some(prev_state) => prev_state,
+            None => return Self::new(config).await,
+        };
+
+        let epoch = Some(prev_state.epoch_params);
+        let msr_enforcer = MSRModeEnforcer::from_preset(
+            config.optimizations.msr_enabled,
```
Comment (raftedproc) [inline ccp/src/prover.rs:130]: IMHO we should store/read MSR opt state from the saved state.
Consider the example, CCP process crashed with  msr enabled, but right before the crash MSR opt has been disabled in the config. The next CCP start(after the failure) will cause the original MSR values to be lost b/c the original preset won't be restored.

**Review by raftedproc — CHANGES_REQUESTED** [review]

File: crates/shared/src/types/epoch_parameters.rs (hunk 5 above)
Comment (mikevoronov) [reply crates/shared/src/types/epoch_parameters.rs:40]: I use it inside `validate_proofs` to create `EpochParameres` in one-line from `CCProofId`

File: ccp/src/proof_storage.rs (hunk 6 above)
Comment (mikevoronov) [reply ccp/src/proof_storage.rs:84]: otherwise formatting will be not the best

File: ccp/src/prover.rs (hunk 7 above)
Comment (mikevoronov) [reply ccp/src/prover.rs:130]: sounds good, will add this
